### ISSUE TITLE
Editor is very laggy for certain actions

### ISSUE THREAD

**Issue by laeubi** [eclipse/lemminx-maven#288]
> M2E - POM Editor using LemMinX language server (includes Incubating components)	1.18.4.20220127-1634	org.eclipse.m2e.lemminx.feature.feature.group	Eclipse.org - m2e

I noticed considerable 'lag' with the new editor for actions that should actually be very fast, in the progress view I see "loading referenced grammar" sometimes.

These action include:

- STRG+D (remove single line)
- STRG+V (insert some code from clipboard)
- STRG+Z (Undo last edit)
- STRG+X

while they run sometimes fast, ther ist often some lag if using them in repetiotn, e.g cut some text in a pom.xml and insert it in other places, also Moving text up/down sometimes lags hard.

It seems as if there is something that has to be started up (and maybe shut down after a while).

The best to reproduce this is the following:

1. close all editors
2. open a pom, as soon as you see the editor focus it and press STRG + D to remove a line
3. It takes on my system around 3-5 seconds until the line is removed and the editor seems unresponsive
4. Press STRG+S to save the change (this also seems to take very long then until the dirty marker disappears
5. Press  STRG+Z ... again a lag of several seconds

I'm using this shortcuts very often and it always has happens fast before this makes the editor feeling very laggy and hard to use.

**Comment by laeubi** [comment]
Also marking some text and the overwrite from clipboard (STRG+V) is very slow.

**Comment by laeubi** [comment]
> Which operating system are you currently using where you are experiencing this bug?

OS is linux debian 64 bit

> Does this issue only occur once you first open up a pom.xml

Nope it also happens afterwards, sometimes an action is fast after triggering it once but gets slow later again (maybe something get wiped out by the GC?).
It also happens if I e.g. copy from one pom a snippet to another pom, then the copy sometimes get lost and some old clipboard content gets inserted.

This doesn't happen for example in the java editor and hasn't happen with the "old" xml editor.

**Comment by mickaelistria** [comment]
Can you pleaase run `jstack` when editor is stuck/laggy?
I don't know where "loading referenced grammar" comes from.

**Comment by laeubi** [comment]
> I don't know where "loading referenced grammar" comes from.

Its probably unrelated ...

> Can you pleaase run `jstack` when editor is stuck/laggy?

I can try that but I'm not sure it will catch the right things as its quite hard to guess *when* it actually happens and is only a few seconds.

**Comment by mickaelistria** [comment]
Isn't it logged in the "Errors" view as a UI Freeze?

**Comment by laeubi** [comment]
No, I don't think this is really an UI freeze but something happening in the back-ground, e.g. if I delete a line with STRG+D and perform this twice there is often a lag of 3-5 seconds and the both lines disappear together, then I can delete some more lines without a problem and after a while (e.g. I mark some text and try to replace it with STRG+V) theres a lag again, thus I think there might a a SoftReference or something of a heavy object getting cleared out by the GC or something alike...

**Comment by laeubi** [comment]
Still seeing the problem in latest 2023-06 Eclipse Release :-1:
